### DISCUSSION TITLE
What is the 'extra' field in Dataset's used for?

### DISCUSSION THREAD

**Discussion by pedro-cf** [apache/airflow#40856] [Q&A] [answered]
What is the 'extra' field in Dataset's used for?

**Comment by jedcunningham** [comment]
It's a free form dictionary - you can use it for whatever you want!

**Reply by pedro-cf** [reply]
> It's a free form dictionary - you can use it for whatever you want!

Is it meant to be used to transfer data from the DAG that triggers the dataset and the DAG that is triggered by the dataset ?

**Reply by jedcunningham** [reply, marked as answer]
No. The extra on Dataset is more meant for static metadata about the dataset. Say, extra identity info or something, like maybe the team that owns it.

There is also an extra on the Dataset events (this is what the scheduling actually happens on). Coming soon in 2.10 is the ability to set the extra on dataset events. That's the right place to pass additional info between the upsteam and downstream sides. You can preview it on our [dev data-aware scheduling docs](http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/stable/authoring-and-scheduling/datasets.html#attaching-extra-information-to-an-emitting-dataset-event).

**Reply by pedro-cf** [reply]
> No. The extra on Dataset is more meant for static metadata about the dataset. Say, extra identity info or something, like maybe the team that owns it.
> 
> There is also an extra on the Dataset events (this is what the scheduling actually happens on). Coming soon in 2.10 is the ability to set the extra on dataset events. That's the right place to pass additional info between the upsteam and downstream sides. You can preview it on our [dev data-aware scheduling docs](http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/stable/authoring-and-scheduling/datasets.html#attaching-extra-information-to-an-emitting-dataset-event).

That's great to hear, meanwhile I've been using this function for this case using task's `triggering_dataset_events`:

```python
def get_triggering_datasets_tasks_xcoms(triggering_dataset_events, dataset_uri):
    logging.info(f"{triggering_dataset_events=}")
    dataset_events = triggering_dataset_events.get(dataset_uri, [])
    logging.info(f"{dataset_events=}")
    xcom_values = []
    for dataset_event in dataset_events:
        xcom_value = DagRun(dag_id=dataset_event.source_dag_id, run_id=dataset_event.source_run_id).get_task_instance(dataset_event.source_task_id).xcom_pull()
        xcom_values.append(xcom_value)
        logging.info(f"Received {xcom_value=} from {dataset_event.source_dag_id=} {dataset_event.source_run_id=} {dataset_event.source_task_id=}")
    return xcom_values
```

**Reply by jedcunningham** [reply]
Yep, xcom is a good workaround in the meantime 🍺
